### ISSUE TITLE
Support Returns for RF enabled systems

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2295]
_Imported from trac ticket 2295.
Created by **kbower**
Opened in trac: Apr 19, 2013, 1:15:08 PM
Last modified in trac: May 27, 2014, 11:05:32 AM_
***
See also #1037

***
* kbower attached [RFID inventorytrans](https://trac.retailarchitects.com/trac/attachment/ticket/2295/RFID inventorytrans) on May 9, 2014, 9:05:20 AM
* kbower attached [return scan same line to more than one location.txt](https://trac.retailarchitects.com/trac/attachment/ticket/2295/return scan same line to more than one location.txt) on May 21, 2014, 9:45:07 AM

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 19, 2013, 1:20:33 PM:_

 * ~~INV_XREF _creation_ at line creation in status PEN.~~ We will only implement this if requested, as [#2329](2329)
 * ~~Line void delete RFID for bulk~~

 * RF OOC Qty
 * ~~Writing returns against open order with RF enabled~~ no demand to implement currently, [#2329](2329)
 * ~~Writing returns for more than was originally sold with RF enabled~~ no demand to implement currently, [#2329](2329)
 * Closing returns with RF
   * Require all RF scanned for RF enabled sites

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 19, 2013, 1:28:30 PM:_

**reply from Mark Bracey inline in bold**:
I need information concerning what needs to occur if the system/store which accepts a return is RF enabled.

Does INV_XREF creation occur at the time the CRM is written **YES. if the ITEM is indicated as Non-Bulk in ITEM.   If the ITEM is a BULK, no INV_XREF is created**(I believe the status is 'PEN' **YES**) or does that occur later **NO**?  Where are the RFIDs found?  ** Currently, if the original order number is referenced on the CRM, the RFID for that SKU and line number referenced is taken out of INV_XREF_ARC and moved to INV_XREF as PEN.  COEM allows for CRM not to reference an original order number, but I don’t know where that RFID comes from.  There should never be a reason to NOT reference am original order number, but that GERS** Are new ones generated or is the attempt made to find the RFID that was shipped?

What needs to occur before the CRM is finalized? **In the Returns department, they are required to Scan the return RFID in the RFCI screen.  If they don’t it will not allow them to finalize the CRM.** Does this depend on whether the PU_DEL_STORE_CD is an RF enabled site, I assume? **~~It is only required if the PU_DEL_STORE_CD is an RF store.~~**  _[**Edit:** this answer was incorrect, see comment:14]_  **If the item is a BULK type item, they have to scan the SKU and then enter the document number.**

Anything else?** Nothing I can think of off the top of my head other than the status change of the XREF record depending on if it is finalized into NAS or regular stock based on how they scanned the RFID in at check in.**

Thanks (I'll need to know this before Returns are supported for Morris),
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 22, 2013, 11:06:02 AM:_

_reply from Mark in _ **bold**
___
Mark: thanks for your answers. More questions:

So, closing an order moves inv_xref to inv_xref_arc? **YES** 

If more than 1 was sold, how does the system decide which RFID? Random? **I tried figuring that out at one point and did determine it was just random.**

When the CRM is written, is the new inv_xref doc_num updated to reference the CRM instead of the SAL? **YES** 

Apparently, Gers allows a CRM to be written against an open ticket with the correct security. In this case, what happens with the inv_xref?  It can't exist twice, once for the CRM and once for the order. **I will have to look into this one and test it first.**

What data prevents the CRM from being closed (how does the system know they haven't all been scanned?) **The PEN INV_XREF record is updated based on how each piece is scanned.  The Store and Location is updated as it is scanned and these are the required fields to enable it to be finalized.  Also, at the time of scanning  the user can scan it into NAS and it [is] assigned an NAS ID.**  You want Loft to prevent the same, correct? **YES ** The inv_xref status?  **The status remains the same (PEN) until it is finalized then either goes into NAS or Available inventory** Or the CRM lines?  Are the CRM lines updated with the location when the scan occurs or not until finalization? **The credit memo lines are updated when it is finalized.** _[Bulk are apparently updated at scan time](Edit:)_


Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 22, 2013, 11:09:14 AM:_

Replying to [kbower](comment:3):
> ''reply from Mark in '' '''bold'''
> Apparently, Gers allows a CRM to be written against an open ticket with the correct security. In this case, what happens with the inv_xref?  It can't exist twice, once for the CRM and once for the order. '''I will have to look into this one and test it first.'''


Until/unless Mark gets back to us we can leave this unimplemented and raise an exception so if a company ever does use this, they can be coerced into doing the testing. ;)

**Comment by kentbower** [comment]
_**sjohnson** trac comment on May 10, 2013, 11:25:45 AM:_

From Johnathan Phelps: 
Writing returns against open order with RF enabled (#2295) – I cannot speak for how other companies do things, but we do not want returns against open orders period. Our thought is that if an order is open it should still be in the store/warehouse, in which case we would just void the piece. If it is open and NOT at our location (i.e. the customer has it), then it needs to be finalized to reflect this and then a return can be done.

Writing returns for more than was originally sold with RF enabled (#2295) – We do not want to give people more money than what they paid for originally.

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 21, 2014, 11:53:22 AM:_

Mark,
There aren't a great number of these, but there are a handful of INV_XREF
records with DISPOSITION 'PEN' for an open CRM for BULK products.

Are those normal?  Currently, when such a Return line is voided, Loft does
not know to delete these (Loft assumes it only needs to look at BULK
products), so it leaves the record there.

If these are generated as part of the normal flow by GERS then Loft should
be deleting them when the Return line is voided, agreed?  So the question
is: are these records flukes or normal?

Thanks,
Kent
___
_from Mark:_
**To be honest, I have never paid attention to the BULK XREF records.  But yes, they should be voided when the line or Order is voided.**

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 21, 2014, 1:17:24 PM:_

(Changeset [[5318]](../commit/cb89bc664c57b73850d7ddac86b3618edb37dd22))
* re #2295, if bulk product Return line is voided, remove any existing bulk INV_XREF records #mor

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 21, 2014, 1:20:42 PM:_

Mark,
I'm trying to implement the closing of Returns for Morris.

   - In this email from a year ago, you mentioned that "at the time of
   scanning  the user can scan it into NAS and it assigned an NAS ID."
    However, you also mentioned "The status remains the same (PEN) until it
   is finalized then either goes into NAS or Available inventory."  So, my
   question is: How does GERS know at finalization time that a certain
   inventory is to be put to NAS if the status (disposition) is kept at PEN? **The disposition will remain as PEN until finalized.  Upon Finalization, it changes the disposition to NAS because of the NAS_OUT_CD entered at the time of the scan upon receipt and the NAS_OUT_ID was populated upon save at the time of scanning.  Also upon finalization the DEL_DOC_NUM and DEL_DOC_LN# is removed.**
   - Does scanning the last item automatically close the Return or must
   someone still use a GERS terminal to close it? **Someone must finalize it in GERS. _[snip]_ **
   - Could you create a Return on TRAIN that is ready to be finalized (has
   had all items scanned, including at least one BULK item, at least one
   scanned into NAS, and at least one going to Available) so I can inspect,
   please? **I entered CRM 0421400WT75 and scanned with all three scenarios**

Thanks!
Kent


On Mon, Apr 22, 2013 at 10:23 AM, Mark Bracey <msb@morrishomefurnishings.com
___
So, how does GERS know whether the NAS_OUT_ID/NAS_OUT_CD refer to NAS or OUT? **I “believe” it is based on the location type codes**

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 22, 2014, 2:47:08 PM:_

Kent/**Mark**:

Does GERS allow the user who is finalizing the CRM to override the scanned location?  What about NAS/OUT code the scanner input?  **Nothing can be changed once it has been scanned with a location and NAS code.  A user will get the error that there is an item pending for that line and con not be changed.  As a side note, if the scanned piece is scanned to a different location before the CRM is finalized, the pending XREF record is updated until it is finalized.**

If the user finalizing the CRM changes the store to one that is not RF enabled, is no scan required? ** As long as the piece has not been scanned into an RF store location, any user can finalize the CRM to any non RF store location as long as the order is not assigned to a truck.**  Same question if changed to store that is RF enabled?

**I’m assuming that these questions are coming based on the PU/DEL store on the order?  if so, the system pays no attention to that field. **

**Comment by kentbower** [comment]
_**kbower** trac comment on May 7, 2014, 4:20:21 PM:_

SQL> update inv_xref set store_cd = '01', loc_cd = 'RETSTG', DEL_DOC_NUM = null, DEL_DOC_LN# = null, DISPOSITION = 'FIF', OOC = 'Y' where del_doc_num =  '0507400WV83';

1 row updated.

Also:
 * ORIGIN_CD
 * NAS_OUT_ID
 * NAS_OUT_CD

___
Set to _something_ even if NULL

**Comment by kentbower** [comment]
_**kbower** trac comment on May 7, 2014, 4:26:39 PM:_

Fix: need to copy OOC flag when archiving and unarchiving.  **_done._**

**Comment by kentbower** [comment]
_**kbower** trac comment on May 9, 2014, 8:21:44 AM:_

Loft Group,
I'm trying to make sense of the RF_ID columns in ITM_TRN during the
finalization of a Return (CMF, OOC, UNR transactions).
It seems terribly arbitrary, almost random as to whether the RF_ID_NEW
actually gets populated with an RF_ID or not.

The best I can tell:

   - CMF is never populated with RF_ID
   - UNR and OOC are only populated if the line qty is exactly 1, it seems?
   - OOC's qty seems to always be 1 even when that isn't line qty (huh?)

Inconsistent data like this is in some cases worse than no data because it
isn't trustworthy, so I am wondering whether Loft should bother with the
RF_ID in ITM_TRN?  If so, it means the qty will always need to be 1 unless
the product is bulk.  So, instead of 1 ITM_TRN of QTY 6, you'd have
6 ITM_TRNs with each QTY 1 and an RFID in RF_ID_NEW.

That would mean more ITM_TRN records, but they'd be consistent.  The thing
I don't want to do in this case is mimic GERS's inconsistent behavior.  I'd
either want to populate RF_ID_NEW consistently or not at all for Return
finalization.

What is your input?

Thanks,
Kent
___
_Mark: (They want RFIDs)_
Funny you brought that up as a GERS inconsistency when it comes to RFIDs in ITM_TRN.  Before we decided to go to V12, we were excited that we could do a query on RFID within the ITQ screen.  None the less, there was a bug in that program and never got fixed for us to use that functionality.  Actually the function is still there and does not work and we, as a user, don’t visibly see that column in the ITQ screen.
If we could clean up ITM_TRN and only give us what we need, that would be great.  Speaking mainly only about CRM finalization right now, I don’t see a reason to be able see OOC, or UNR when CMF is the only thing we need to see.  Then the RFID could live on that line.
OOC really never needs to be in ITM_TRN.

**Comment by kentbower** [comment]
_**kbower** trac comment on May 12, 2014, 8:30:12 AM:_

**Mark:**
Sounds like I need to retract my comment [made previously]. 
There should not be a cross check based on PU/DEL store.  In order to finalize the CRM to an RF store, the item must be scanned first.  If something is already scanned into an RF store, no changes can be made to those lines.  If all lines on the CRM are scanned, anyone can finalize the CRM but no matter what, it cannot be finalized to any locations other than  current scanned location.
___
**Kent:**
Right, but how does it know they need to be all scanned?  It needs to check If RF is enabled for some store. If not PU/DEL store, then which?  Or does it not allow you to choose a site that is RF enabled when you go to fill in STORE_CD while finalizing CRM?
___
**Mark:**
**It does not allow you to choose a store CD that is RF enabled when finalizing.**

**Comment by kentbower** [comment]
_**kbower** trac comment on May 12, 2014, 2:47:48 PM:_

RFID retains outlet status and id when unarchived at write of return (until scanned):
```
RF_ID_CD   STOR LOC_CD ITM_CD    DEL_DOC_NUM    DEL_DOC_LN# SE_PART_SEQ IST_WR_DT            IST_ IST_SE    IST_LN# NAS_OUT_ID NAS CRPT_ID_NU CREATE_DT            LAST_ACTN_DT         LAST_CYCLE_DT        LAST_ACTN_EMP_C ACT DIS        QTY ORIGI REPLACED_B ORI SER_NUM                     CST O
---------- ---- ------ --------- -------------- ----------- ----------- -------------------- ---- ------ ---------- ---------- --- ---------- -------------------- -------------------- -------------------- --------------- --- --- ---------- ----- ---------- --- -------------------- ---------- -
0019506991 02   OUTLET 425266776 1110202THOI              1                                                         0825264945 OUT            12-JUL-2012 00:00:00 10-NOV-2012 00:00:00                      4107                FIN          1 C1                                                   N

RF_ID_CD   STOR LOC_CD ITM_CD    DEL_DOC_NUM    DEL_DOC_LN# SE_PART_SEQ IST_WR_DT            IST_ IST_SE    IST_LN# NAS_OUT_ID NAS CRPT_ID_NU CREATE_DT            LAST_ACTN_DT         LAST_CYCLE_DT        LAST_ACTN_EMP_C ACT DIS        QTY ORIGI REPLACED_B ORI SER_NUM              T C ST        CST O
---------- ---- ------ --------- -------------- ----------- ----------- -------------------- ---- ------ ---------- ---------- --- ---------- -------------------- -------------------- -------------------- --------------- --- --- ---------- ----- ---------- --- -------------------- - - -- ---------- -
0019506991             425266776 0512402ILLI              1                                                         0825264945 OUT            12-JUL-2012 00:00:00 12-MAY-2014 00:00:00                      6350                PEN          1 SOE                                                         N
```

**Comment by kentbower** [comment]
_**kbower** trac comment on May 21, 2014, 3:21:44 PM:_

## Unanswered by Morris (behavior pieced together, not verified)
I'm glad you made 0513401ILLN because I want to see if it splits the line only on differing locations or also on differing NAS status (with same location).  
For 0513401ILLN line 1, please scan one to AA1 and the other to AA2 (both available).
For line 2, please scan both to AA3, but one as NAS and one available.

Kent
**_(Assume it splits Return line.  Loft will just write the first location to the line, the other qtys can be researched in transactions.)_**
___
>On Tue, May 13, 2014 at 4:41 PM, Mark Bracey <Mark.Bracey@morrisfci.com> wrote:
>If by chance they accidently scanned a piece that did not get returned, someone in IT >has to go into the XREF table (TFIX) to remove the location and NAS code and ID if >applicable before they can remove or split that line off.


What if it was qty 2 line and they already scanned one of the two (the one that was returned)?  Do you still need to TFIX the XREF just so you can drop the line qty from 2 to 1 and then rescan it, or does GERS allow a qty drop when the line has begun to be scanned?

**_(Loft won't allow any changes if even part of the line is scanned.  We may enhance this, but would be nice to know what GERS does.)_**

**Comment by kentbower** [comment]
_**kbower** trac comment on May 27, 2014, 11:05:32 AM:_

(Changeset [[5403]](../commit/80130474b320858504faa5e072adfe534e22495a))
* closes [#2317](2317), Implement closing returns for V12 #mis #mor
 * closes #2295, Support Returns for RF enabled systems #mor

 * restructure/rewrite of closing Returns/Exchanges to support V12 and RF #all

**Comment by kentbower** [comment]
_**kbower** trac status change on May 27, 2014, 11:05:32 AM:_ **closed**
